feat: implement of btc-staker improvement 107

wnjoon opened the pull request:
This PR includes the implementation for the btc-staker improvement related to issue #107 .

The removal of the watch-staking endpoint and the post-approval flow were approved in the previous PR(#131). This PR focuses on 'reducing the amount of state stored in the database'.

The details of the implementation are as follows:

## Reduce the amount of state held in the Database

Previously, btc-staker had the following structural shortcomings due to the inclusion of watch-staking endpoints and post-approval flows:
- It required maintaining status information for both Babylon and Bitcoin, even though certain information could be directly queried from these chains. This led to unnecessary duplication in the database.
- The logic became overly complex, as follow-up processing had to account for both post-approval and pre-approval scenarios when restarting the program.

In the previous PR(#131), the watch-staking endpoint and post-approval flow were removed, simplifying the database structure. Additionally, information that can be directly queried from the chain (e.g., transaction progress status) is no longer stored in the database.

The existing database structure is as follows:
> The `Watched` field is not shown as it was already removed in a previous PR.

```go
type StoredTransaction struct {
    StoredTransactionIdx      uint64
    StakingTx                 *wire.MsgTx
    StakerAddress              string
    StakingOutputIndex        uint32
    StakingTxConfirmationInfo *BtcConfirmationInfo
    StakingTime               uint16
    FinalityProvidersBtcPks   []*btcec.PublicKey
    Pop                       *ProofOfPossession
    State                      proto.TransactionState
    UnbondingTxData            *UnbondingStoreData
    BabylonBTCDelegationTxHash string
}
```

The reduced db looks like this:

```go
type StoredTransaction struct {
	StoredTransactionIdx uint64
	StakingTx            *wire.MsgTx
	StakerAddress        string
}
```

The deleted fields can be retrieved directly from Babylon or Bitcoin when needed. Additionally, some fields are no longer used due to the removal of the post-approval flow.

wnjoon commented:
Commit Summary [0233d1a](https://github.com/babylonlabs-io/btc-staker/pull/134/commits/0233d1af023f34610840f588ff9f69a05882832b)

- Removed state, fpkey, and pop fields from StoredTransaction
- Added a function `QueryBTCDelegation` to query transaction information from the Babylon chain.

wnjoon commented on staker/stakerapp.go:
```diff
@@ -567,354 +423,147 @@ func (app *App) SendPhase1Transaction(
+	if unbondingTxStatus == walletcontroller.TxNotFound {
+		// no unbonding tx on chain and staking output already spent, most probably
+		// staking transaction has been withdrawn, update state in db
+		return nil
```
I think it would be a good idea to remove that transaction content from the database. Otherwise, functions that query the entire database, such as QueryStoredTransactions, will retrieve unnecessary items.

wnjoon replied:
fixed [380b429](https://github.com/babylonlabs-io/btc-staker/pull/134/commits/380b4292df4e0f2a1c64275d83522b86b77bd2bc)

wnjoon commented on staker/stakerapp.go:
```diff
@@ -2206,8 +1514,23 @@ func (app *App) SpendStake(stakingTxHash *chainhash.Hash) (*chainhash.Hash, *btc
+		return nil, nil, fmt.Errorf("cannot spend staking output. Error converting fpBtcPkList to btcPkList: %w", err)
+	}
+
 	spendStakeTxInfo, err := createSpendStakeTxFromStoredTx(
```
I modified the createSpendStakeTxFromStoredTx function and its calling part to align with your suggestion.

We also need to discuss the WithdrawableTransactions function.

Currently, this function queries all transactions stored in the database and checks whether `StakingTxConfirmedOnBtc()` or `UnbondingTxConfirmedOnBtc()` is set. Based on this, it sets the corresponding values and then verifies whether `timeLockExpired` is used.

Due to this structure, we had to maintain `StakingTxConfirmationInfo` and `UnbondingStoreData.UnbondingTxConfirmationInfo` in the database. Additionally, we also store the `StakingTime` and `UnbondingTime` fields in the database.

If these fields are removed, we would need to call the RPC for each transaction to fetch the required values, which seems highly inefficient. Moreover, adding an RPC call inside the `QueryStoredTransactions` function does not seem structurally ideal.

If querying all withdrawable transactions is not necessary, it might be a better approach to implement a function that checks whether specific requested transactions are withdrawable. Please share your thoughts on this.

wnjoon replied:
We need to discuss the WithdrawableTransactions function.

Currently, this function queries all transactions stored in the database and checks whether StakingTxConfirmedOnBtc() or UnbondingTxConfirmedOnBtc() is used. Based on this, it sets the corresponding values and then verifies whether timeLockExpired is used.

Due to this structure, we had to maintain StakingTxConfirmationInfo and UnbondingStoreData.UnbondingTxConfirmationInfo in the database. Additionally, we also store the StakingTime and UnbondingTime fields in the database.

If these fields are removed, we would need to call the RPC for each transaction to fetch the required values, which seems highly inefficient. Moreover, adding an RPC call inside the QueryStoredTransactions function does not seem structurally ideal.

If querying all withdrawable transactions is not necessary, it might be a better approach to implement a function that checks whether specific requested transactions are withdrawable. Please share your thoughts on this.

wnjoon replied:
> this functions has inside the follwing checks:
> 
> ```
> if storedtx.StakingTxConfirmedOnBtc() && !storedtx.UnbondingTxConfirmedOnBtc() {
> ```
> 
> and `storedtx.StakingTxConfirmedOnBtc() ` was depended on the old post approval flow.
> 
> I think now we must remove this check.
> 
> This also touches:
> 
> ```
> func (app *App) WithdrawableTransactions(limit, offset uint64) (*stakerdb.StoredTransactionQueryResult, error) {
> ```
> 
> function i.e this function currently returns results based on this field.
> 
> I think now delegation is withdrawable if:
> 
> * the staking transaction is still on BTC and its timelock has expired
> * the unbonding transaction is still on BTC and its timelock has expired

We need to discuss the WithdrawableTransactions function.

Currently, this function queries all transactions stored in the database and checks whether StakingTxConfirmedOnBtc() or UnbondingTxConfirmedOnBtc() is used. Based on this, it sets the corresponding values and then verifies whether timeLockExpired is used.

Due to this structure, we had to maintain StakingTxConfirmationInfo and UnbondingStoreData.UnbondingTxConfirmationInfo in the database. Additionally, we also store the StakingTime and UnbondingTime fields in the database.

If these fields are removed, we would need to call the RPC for each transaction to fetch the required values, which seems highly inefficient. Moreover, adding an RPC call inside the QueryStoredTransactions function does not seem structurally ideal.

If querying all withdrawable transactions is not necessary, it might be a better approach to implement a function that checks whether specific requested transactions are withdrawable. Please share your thoughts on this.

KonradStaniec replied:
>If querying all withdrawable transactions is not necessary, it might be a better approach to implement a function that checks whether specific requested transactions are withdrawable. Please share your thoughts on this.

Tbh I feel this function is hard requiremnt for the users i.e If I am a user and have multiple delegations (each in different stage) I should have way to check which of those I can withdraw to my wallet.

In the future such functionality can be automatic based on this information i.e
- periodically query all withdrawable transactions
- and withdraw them to user wallet

> If these fields are removed, we would need to call the RPC for each transaction to fetch the required values, which seems highly inefficient. Moreover, adding an RPC call inside the QueryStoredTransactions function does not seem structurally ideal.

ahh agreed on that. this means we need to fill `StakingTxConfirmedOnBtc()` field in db, after tx delegation is active on Babylon

wnjoon replied:
fixed [cb94a22](https://github.com/babylonlabs-io/btc-staker/pull/134/commits/cb94a2207a77cc921d6277de6ec6160afcc1800b)

@KonradStaniec 
I modified `WithdrawableTransaction` to retrieve the necessary data via RPC instead of the database.
As a result, since the database is no longer used, I created it as a separate function rather than passing a query statement to `QueryStoredTransactions`. After retrieving the status from the Babylon chain, I then obtained the confirmation information from Bitcoin based on the transaction status.

wnjoon commented on staker/stakerapp.go:
```diff
@@ -1171,12 +824,13 @@ func (app *App) sendUnbondingTxToBtc(
 	defer app.wg.Done()
 	defer waitEv.Cancel()
-	unbondingTxHash := unbondingData.UnbondingTx.TxHash()
+	// unbondingTxHash := unbondingData.UnbondingTx.TxHash()
```
fixed [380b429](https://github.com/babylonlabs-io/btc-staker/pull/134/commits/380b4292df4e0f2a1c64275d83522b86b77bd2bc)

wnjoon commented on stakerservice/service.go:
```diff
@@ -217,15 +210,21 @@ func (s *StakerService) stakingDetails(
 	}
 
 	details := storedTxToStakingDetails(storedTx)
+	details.StakingState = di.BtcDelegation.GetStatusDesc()
```
fixed [380b429](https://github.com/babylonlabs-io/btc-staker/pull/134/commits/380b4292df4e0f2a1c64275d83522b86b77bd2bc)

wnjoon commented:
Comment Summary [380b429](https://github.com/babylonlabs-io/btc-staker/pull/134/commits/380b4292df4e0f2a1c64275d83522b86b77bd2bc)

- Removed db fields: StakingOutputIndex, BabylonBTCDelegationTxHash
- Refactoring db field: Take out UnbondingTxConfirmationInfo from UnbondingTxData
- Added a function to remove transactions from the database and applied it to the logic where a transaction is spent.

wnjoon commented:
Comment Summary [cb94a22](https://github.com/babylonlabs-io/btc-staker/pull/134/commits/cb94a2207a77cc921d6277de6ec6160afcc1800b)

- Removed db fields related to confirmation
- Changed the logic of the WithdrawableTransaction function to retrieve data via RPC instead of the database.

At this stage, the database will look like this:

```go
type StoredTransaction struct {
	StoredTransactionIdx uint64
	StakingTx            *wire.MsgTx
	StakerAddress        string
}
```

wnjoon commented:
Commit Summary [8628a24](https://github.com/babylonlabs-io/btc-staker/pull/134/commits/8628a246953f11a4aee8dd63eda4900c5c5cea12)

- Updated protobuf
- Removed unused functions
- Added comments
- Wrapped errors using `fmt.Errorf` based on the code I worked on

KonradStaniec commented on staker/stakerapp.go:
```diff
@@ -2046,34 +1374,110 @@ func (app *App) StoredTransactions(limit, offset uint64) (*stakerdb.StoredTransa
+
+	var totalTxCount int
+	// add stored transactions to slice
+	if err := app.txTracker.ScanTrackedTransactions(func(tx *stakerdb.StoredTransaction) error {
```
I think here we need to:
- first return all the hashes in db (similarity like in `checkTransactionsStatus`)
- and then iterate over all the hashes to check the status  on Babylon and BTC

Otherwise we will take the read lock on db for a long time given all the queries we are making.

wnjoon replied:
You are right.
I will try modifying the function by applying offset and limit as described below.

Previously, I returned storedTransaction from ScanTrackedTransactions and processed it directly in RPC, so I didn’t feel the need for offset and limit.

I will change the logic as follows:
1. Apply offset and limit to the StoredTransactions function to first retrieve the transaction list.
2. Process the retrieved transactions one by one based on their status.

KonradStaniec commented on proto/transaction.proto:
```diff
@@ -4,68 +4,15 @@ package proto;
     bytes block_hash = 2;
 }
 
-message CovenantSig {
```
I think `BTCConfirmationInfo` is not longer used

wnjoon replied:
fixed [2c2adee](https://github.com/babylonlabs-io/btc-staker/pull/134/commits/2c2adeee1540dec749fec0c67c0759b06a8f4720)

KonradStaniec commented on stakerservice/service.go:
```diff
@@ -401,249 +416,39 @@ func (s *StakerService) withdrawableTransactions(_ *rpctypes.Context, offset, li
 		"list_staking_transactions":          rpc.NewRPCFunc(s.listStakingTransactions, "offset,limit"),
 		"unbond_staking":                     rpc.NewRPCFunc(s.unbondStaking, "stakingTxHash"),
-		"withdrawable_transactions":          rpc.NewRPCFunc(s.withdrawableTransactions, "offset,limit"),
+		"withdrawable_transactions":          rpc.NewRPCFunc(s.withdrawableTransactions, ""),
```
why do you think pagination does not make sense here ?

wnjoon replied:
fixed [2c2adee](https://github.com/babylonlabs-io/btc-staker/pull/134/commits/2c2adeee1540dec749fec0c67c0759b06a8f4720)

KonradStaniec reviewed (approved):
Looks good !  thanks for this contribution !
